src/libsystemd/sd-netlink/nfnl-message.c

```diff
@@ -0,0 +1,323 @@
+                                        type);
+        if (r < 0)
+                return r;
+
+        m->hdr->nlmsg_len = size;
+        m->hdr->nlmsg_type = NFNL_SUBSYS_NFTABLES << 8 | type;
+
+        nfh = NLMSG_DATA(m->hdr);
+        nfh->nfgen_family = AF_INET;
+        nfh->version = NFNETLINK_V0;
+        nfh->res_id = nfnl->serial;
+
+        *ret = TAKE_PTR(m);
+
+        return 0;
+}
+
+static int sd_nfnl_message_batch(sd_netlink *nfnl, sd_netlink_message **ret, int v)
+{
+        struct nfgenmsg *nfh;
+        int r;
+
+        r = message_new(nfnl, ret, v);
+        if (r < 0)
+                return r;
+
+        nfh = NLMSG_DATA((*ret)->hdr);
+        nfh->nfgen_family = AF_UNSPEC;
+        nfh->version = NFNETLINK_V0;
+        nfh->res_id = NFNL_SUBSYS_NFTABLES;
+
+        return r;
+}
+
+int sd_nfnl_message_batch_begin(sd_netlink *nfnl, sd_netlink_message **ret)
+{
+        return sd_nfnl_message_batch(nfnl, ret, NFNL_MSG_BATCH_BEGIN);
+}
+
+int sd_nfnl_message_batch_end(sd_netlink *nfnl, sd_netlink_message **ret)
+{
+        return sd_nfnl_message_batch(nfnl, ret, NFNL_MSG_BATCH_END);
+}
+
+int sd_nfnl_nft_message_new_basechain(sd_netlink *nfnl, sd_netlink_message **ret,
+                                      const char *table, const char *chain,
+                                      const char *type,
+                                      uint8_t hook, int prio) {
+        sd_netlink_message *m;
+        int r;
+
+        r = nft_message_new(nfnl, ret, NFT_MSG_NEWCHAIN, NLM_F_ACK);
+        if (r < 0)
+                return r;
+
+        m = *ret;
+
+        r = sd_netlink_message_append_string(m, NFTA_CHAIN_TABLE, table);
+        if (r < 0)
+                return r;
+
+        r = sd_netlink_message_append_string(m, NFTA_CHAIN_NAME, chain);
+        if (r < 0)
+                return r;
+
+        r = sd_netlink_message_append_string(m, NFTA_CHAIN_TYPE, type);
+        if (r < 0)
+                return r;
+
+        r = sd_netlink_message_open_container(m, NFTA_CHAIN_HOOK);
+        if (r < 0)
+                goto cancel;
+
+        r = sd_netlink_message_append_u32(m, NFTA_HOOK_HOOKNUM, htobe32(hook));
+        if (r < 0)
+                goto cancel;
+
+        r = sd_netlink_message_append_u32(m, NFTA_HOOK_PRIORITY, htobe32(prio));
+        if (r < 0)
+                goto cancel;
+
+        r = sd_netlink_message_close_container(m);
+        if (r < 0)
+                goto cancel;
+
+        return 0;
+cancel:
+        sd_netlink_message_cancel_array(m);
+        return r;
+}
+
+int sd_nfnl_nft_message_del_table(sd_netlink *nfnl, sd_netlink_message **ret,
+                                  const char *table)
+{
+        int r = nft_message_new(nfnl, ret, NFT_MSG_DELTABLE, NLM_F_ACK);
+        if (r < 0)
+                return r;
+
+        r = sd_netlink_message_append_string(*ret, NFTA_TABLE_NAME, table);
+        if (r < 0)
+                return r;
+
+        return r;
```

Comment:
as above: on success initialize all return values and on failures none. 

(here and everywhere else)